README.md
<img src=http://www.nyit.edu/files/communications_and_marketing/DIGITAL_LOGO_Engineering_RGB_HORIZ.png width="179" height="63" />

<br>

# Indoor Localization Framework with WiFi Fingerprinting

<p align="center">
  <img src=https://rkhullar.github.io/csci870/images/icon.png width="256" height="256" />
</p>

<br>
<br>

| Project Member  | **Rajan Khullar** |
| --------------- | ----------------- |
| Project Advisor | Dr. Ziqian Dong   |
| Semester        | Fall 2016         |

<div style="page-break-after: always;"></div>

## Introduction
### Abstract
As with most studies, the results of WiFi fingerprinting are much more meaningful if there is a large sample size. For this project I created an Android app that helps researchers in this field build their dataset  efficiently. After collecting two weeks worth of data I studied how many access points are ideal to predict a user's location from a single WiFi scan.

### Background
Global Positioning Systems (GPS) have been the standard technology used to obtain location of electronic devices. GPS devices are essential for the functionality of many of today’s devices, appearing in vehicles, drones, and smart phones. They have a wide range of applications, from helping people navigate the roads to assisting the military in navigating planes and drones. Moreover, the popularity of smartphones with their inbuilt GPS devices has made GPS easily available.

Unfortunately localization using GPS is ineffective indoors and in highly metropolitan environments because of GPS signal fading. Signal fading occurs when signals penetrate building materials causing a decrease in intensity. This causes a decrease in signal-to-noise ratio making it difficult for GPS devices to differentiate between signals and noise. Signal fading is also a result of multipath phenomenon, which is caused by reflection and refraction of signals when they encounter walls [1]. Because of these reasons GPS signal is often lost entirely on smartphones in indoor environments, causing a pressing need for indoor localization.

Indoor Localization may be utilized similarly to outdoor localization. This includes commercial applications such as real-time maps to help people navigate within malls or museums, and more technical applications such as guiding drones through indoor environments or locating cars inside tunnels. An indoor positioning system will open a new market of applications.

The most accurate way to express WiFi signal strength is with dBm, which stands for decibels relative to one milliwatt. Another common meaurement is the Received Signal Strength Indicator (RSSI). The RSSI measures the relative quality of the signal and is expressed with dB. The Android API provides signal strength as the latter. [5]

| $$\textrm{dBm}=10\times\log{\dfrac{P}{1\quad\textrm{mW}}}$$ | $$\textrm{dB}=\log{\dfrac{P_1}{P_2}}$$ |
| :--------------------------------------: | :------------------------------------: |
|           ***Absolute Power***           |         ***Ratio of Powers***          |

<div style="page-break-after: always;"></div>

|        ![REU 2013][reu13]         |         ![REU 2015][reu15]         |
| :-------------------------------: | :--------------------------------: |
| ***Figure 1.1 - EGGC 6th Floor*** | ***Figure 1.2 - MC16 Auditorium*** |

### Related Work
Indoor Localization has been attempted through the use of Wi-Fi signal strength and Sensor fusion. Kothari et al (2012) [2] have successfully used dead reckoning and Wi-Fi signal strength fingerprinting to find the location of a smartphone. Dead reckoning was their method of using the accelerometer, gyroscope, compass and a particle filter in order to track walking and thereby track location. Both of these methods are prone to large errors. Wi-Fi signal strength is affected by obstacles and a myriad of other Wi-Fi signals in an urban environment, while the accelerometer and gyroscope sensors are likely to generate random noise in the data.

Thanks to the National Science Foundation's Research Experience Undergraduate (REU) program, research fellows have been able to study indoor localization at NYIT. As shown in Figure 1.1, students in the summer of 2013 chose three access points on a single floor and measured signal strengths from twenty six spots evenly distributed in the hallways. [3]

In Summer 2015 a student from Cooper Union and I were two of the REU fellows. We attempted to correlate energy consumption of a Nexus 5 to physical distance from an access point. We setup one router as a fixed access point and designed an Android application that records the current and voltage of the phone while pinging the router. We took ten samples at sixty points in a large classroom, however we were unable to find any significant correlation as shown in Figure 1.2. 

In Summer 2016 two REU fellows studied multifloor localization with four consecutive floors in NYIT's main building. They were able to distinguish between thirty locations in their dataset with high accuracy. In all three REU studies the process of data collection proved difficult. The third project had a sample size of around 300 WiFi scans. After realizing this I was inspired to create a framework to help automate the process of gathering samples.

<div style="page-break-after: always;"></div>

## Implementation
|               ![frw][frw]                |
| :--------------------------------------: |
| ***Figure 2.1 - High Level System Design*** |

The Digital Ocean server has Apache and PostgreSQL installed. A python library called Flask was used to create a REST api. Java was used to create the Android application. As shown in Figure 3.1, once users sign up and login they can choose their classroom and setup a scan for the duration of that class. The scans will occur in the background so the users can close the app and use their phone normally. The scans can be paused or canceled in case the users needs to change their location. Finally once the scans are complete, then each user can upload their local dataset to the server.

For my analysis, the table of scan records if first downloaded from my server. Then it is preprocessed before being used to train the location classifiers. Before testing the prediction accuracy of the classification models, we analyze how much data we have in terms of both scan records and complete scans. We are also able to visualize the WiFi fingerprints for each location.

The graphs shown in this report represent only a small portion of all the graphs generated. All the graphs are stored on github with the following URL: [rkhullar.github.io/csci870/][M]

|      ![app-setup][app1]       |          ![app-push][app2]           |
| :---------------------------: | :----------------------------------: |
| ***Figure 3.1 - Setup Scan*** | ***Figure 3.2 - Push Scan Results*** |

### Database

|               ![erd][erd]                |
| :--------------------------------------: |
| ***Figure 2.2 - Entity Relation Diagram*** |

One WiFi scan or sample results in one or more scan records. Each scan record contains the the mac address for a single access point, the signal strength to that access point, the location of the scan, the UNIX time stamp, and the user who performed the scan. The locations are stored as a separate table with building, floor, and room description.

In order to reduce redundancy a table of unique access points is maintained as well as one for unique locations. The actor table contains information for all people in the system including normal users, administrators, and new unverified signups. Figure 2.2 shows the simplified entity relation diagram.

### Preprocessing
|               ![pre][pre]               |
| :-------------------------------------: |
| ***Figure 2.3 - Preprocessing Module*** |

Figure 2.3 shows how the dataset is prepared for machine learning. First the table of scan records is downloaded from the server by an administrator. The program groups each record by location and hour. Groups that do not have at least 1000 records are ignored. Each passing group is further grouped by the WiFi access point into blocks. Each block must contain at least 100 records. Then all the passing access points become columns and the passing records are combined by their timestamps into complete scans. The day of week and hour are also extracted from each timestamp.

#### Input
| UXT  | BSSID | Signal Strength | Location |
| :--: | :---: | :-------------: | :------: |
|  T1  |  W1   |       *W*       |    L1    |
|  T1  |  W2   |       *X*       |    L1    |
|  T2  |  W2   |       *Y*       |    L2    |
|  T2  |  W3   |       *Z*       |    L2    |

#### Output
|   DoW   |   Hour   |  W1  |  W2  |  W3  | Location |
| :-----: | :------: | :--: | :--: | :--: | :------: |
| dow(T1) | hour(T1) | *W*  | *X*  | *-*  |    L1    |
| dow(T2) | hour(T2) | *-*  | *Y*  | *Z*  |    L2    |

### Machine Learning
|   ![svm-linear][svm-linear]   |       ![svm-poly][svm-poly]       |   ![svm-radial][svm-radial]   |
| :---------------------------: | :-------------------------------: | :---------------------------: |
| ***Figure 4.1 - Linear SVM*** | ***Figure 4.2 - Polynomial SVM*** | ***Figure 4.3 - Radial SVM*** |

In this study we use the linear support vector machine classifier to perform location prediction. As shown in Figure 4.1 lines are generated to divide the hyperplane into labeled classes. Figure 4.3 shows an example of radial classification, but this usually leads to overfitting. Another countermeasure to this problem is to use *k* fold cross validation. The entire dataset of samples is randomly divided into *k* evenly sized blocks. Each block is used as testing data once while all the other blocks are used as training data. In this way *k* classifiers are actually trained and tested. [4]

<div style="page-break-after: always;"></div>

## Results and Analysis

### Group Cardinality
|            ![location][cntL]             |
| :--------------------------------------: |
| ***Figure 5.1 - Scan Records grouped by Location*** |

Figure 5.1 shows that 17 locations passed the intial filter of having at least 1000 scan records. The top four locations have over 15,000 records.

<div style="page-break-after: always;"></div>

|              ![hour][cntT]               |
| :--------------------------------------: |
| ***Figure 5.2 - Scan Records grouped by Hour*** |

Figure 5.2 shows that most of the scans were performed between 4 PM and 7 PM.

<div style="page-break-after: always;"></div>

|          ![access point][cntW]           |
| :--------------------------------------: |
| ***Figure 5.3 - Scan Records grouped by Access Point*** |

Forty three access points passed the intitial filter which means there are that many features plus two in total. The most common access point is nearly twice as prevalent as the second most common access point. 

<div style="page-break-after: always;"></div>

|          ![location-wap][cntF]           |
| :--------------------------------------: |
| ***Figure 5.4 - Filtered Scans grouped by Location*** |

Before preprocessing there were 17 locations that passed the initial filter. After preprocessing only 14 locations are available to train the classifiers. Figure 5.4 shows how many scans there are for earch of those 14 locations. L01 through L05 have at least 3000 scans and the rest have under 1500 scans.

<div style="page-break-after: always;"></div>

### Signal Strength
|             ![distX][distX]              |
| :--------------------------------------: |
| ***Figure 6 - Overall Distribution of Signal Strengths*** |

Figure 6 shows that the best and worst signal strength's recorded in my dataset were -20 dB and -95 dB respectively. The signal strength is normally distributed with mean of -70 dB.

<div style="page-break-after: always;"></div>

### Fingerprints
|              ![box1][box1]               |              ![box2][box2]               |
| :--------------------------------------: | :--------------------------------------: |
| ***Figure 7.1 - Fingerprint for EGGC 601 at 10 AM*** | ***Figure 7.2 - Fingerprint for EGGC 601 at 12 PM*** |

|              ![box3][box3]               |              ![box4][box4]               |
| :--------------------------------------: | :--------------------------------------: |
| ***Figure 7.3 - Fingerprint for EGGC 704 at 6 PM*** | ***Figure 7.4 - Fingerprint for EGGC 704 at 7 PM*** |

|            ![dist-WL][distWL]            |
| :--------------------------------------: |
| ***Figure 7.5 - Scan Density per Location and Access Point*** |

Figures 7.1 through 7.4  show that WiFi scan fingerpints for individual locations are similar regardless of time, and that they are different for unique locations. Figure 7.5 shows how many samples are in each location and access point group. Each row can be interpreted as the fingerprint for one location as well.

<div style="page-break-after: always;"></div>

### Prediction Accuracy
|              ![mtx1][mtx1]               |
| :--------------------------------------: |
| ***Figure 8.1 - Performance with All Access Points*** |


|              ![mtx2][mtx2]               |              ![mtx3][mtx3]               |
| :--------------------------------------: | :--------------------------------------: |
| ***Figure 8.2 - Performance with One WAP*** | ***Figure 8.3 - Performance with One WAP and Time*** |

|              ![mtx4][mtx4]               |              ![mtx5][mtx5]               |
| :--------------------------------------: | :--------------------------------------: |
| ***Figure 8.4- Performance with 11 WAPs*** | ***Figure 8.5 - Performance with 11 WAPs and Time*** |

Figures 8.1 through 8.5 compare the location predictors that were trained from my dataset. Since 43 access points passed the inital filter there are a total of 45 features that can be used. Two of them are the day of week and hour of day, which are extracted from the timestamp for a scan. By using all the access points the location can be predicted with 99% accuracy. Most of the false predictions come from classifying scans from L10 as L07.

If only the most common access point is utilized then nearly all of the locations for each scan are predicted incorrectly as L05. The exceptions are samples from L02 and L05 which are still predicted correctly at least 90% of the time. By using the most common access point along with day of week and hour, the prediction accuracy significantly improves from 37% to 84%. By using the top eleven most common access points along with time, the prediction accuracy becomes 100%. Without time the accuracy is still less than when using time with only one access point.

<div style="page-break-after: always;"></div>

|             ![decay][decay]              |
| :--------------------------------------: |
| ***Figure 9 - Prediction Accuracy vs Number of WAP Features*** |

Figure 9 shows how the prediction accuracy changes with the number of access points used. Since forty three waps were used, there are eighty seven classifiers in total. The red line starts by using the most common access point. The green line start by only using the day of week and hour of day. Ideally using no access points should yield an accuracy of $$N^{-1}$$ where $$N$$ represents the number of locations to predict from. That would be the same as randomly choosing one location.

Since the classifier were trained and tested using 10 fold cross validation, there actually 10 accuracies for each one. The figure shows the average and and standard deviation for each classifier. The error bars on the green line are much larger then in the red line, which means that the scores for each individual classifier are much more consistent when only using the access points as features.

<div style="page-break-after: always;"></div>

## Conclusion

Prediction accuracy is significantly improved by including time features in the classification algorithm. With my dataset 100% accuracy can be achieved by using time and 11 access points. Without time the 28 most common access points are required to get close to the same accuracy. However as shown in Figures 5.1 and Figure 5.2, my dataset is highly unbalanced with regard to both time and location. That might be the reason we see such high improvement by using time as a feature.

## Future Work

The Android application should be modified to record the phone’s model number. This is important since the signal reading is dependant on the antenna and each model of phone may have a unique type of antenna. For example if an iphone and a nexus were to perform a WiFi scan at the exact same time and location, the results may not be the same. They should pick up the same access points but the signal strengths to each access point might be offset. 

In order to easily balance the dataset, raspberry pi’s can be placed in the perimeter of each room and be programed to automatically perform WiFi scans at a set time interval. When creating the location classifiers, the scans from the pi's should be used as training data. The scans from the mobile devices can be used as testing data. It would be interesting to see how accurate the predictors would be in that situation.

The signal strengths picked up by each type of mobile device may have a fixed offset to those picked up by the pi's. If that is the case then the training data would not need to come from phones at all and using one kind of device would suffice. Before predicting location with a scan from a phone, the signal strengths received would simply be offset appropriately.

## Learning Outcomes
| Server            | Android              | Machine Learning     |
| ----------------- | -------------------- | -------------------- |
| PostgreSQL        | SQLite               | Training Classifiers |
| REST API in Flask | Broadcast Receivers  | Cross Validation     |
| Apache with HTTPS | Background Services  | Confusion Matrices   |
| Sending Email     | Making HTTP Requests | Matplotlib           |
|                   | Notifications        |                      |

<div style="page-break-after: always;"></div>

## References
1. Kjærgaard et al. "Indoor Positioning Using GPS Revisited," *Proceedings of the 8th International Conference on Pervasive Computing*, Helsinki, Finland. Springer-Verlag, 2010. 38-56
2. N. Kothari, B. Kannan, E. Glasgwow, M. Dias, Robust Indoor Localization on a Commercial Smart Phone. *Procedia Computer Science*,  2012. 1114-1120
3. N. Gutierrez, C. Belmonte, J. Hanvey, R. Espejo, Z. Dong, Indoor localization for mobile devices. *Proceedings of the 11th IEEE International Conference on Networking, Sensing and Control*, Miami, FL, 2014. 173-178.
4. [Scikit-learn: Machine Learning in Python][R3], Pedregosa et al., JMLR 12, pp. 2825-2830, 2011.
5. DB vs. DBm. Editorial. *InTech Magazine*, 1 Nov. 2002. ISA. 

[R1]: http://dx.doi.org/10.1007/978-3-642-12654-3_3
[R2]: http://www.sciencedirect.com/science/article/pii/S1877050912005157
[R3]: http://ieeexplore.ieee.org/document/6819620/
[G4]: http://www.scipy-lectures.org/advanced/scikit-learn/
[R4]: http://jmlr.csail.mit.edu/papers/v12/pedregosa11a.html
[R5]: https://www.isa.org/standards-publications/isa-publications/intech-magazine/2002/november/db-vs-dbm/

[M]: https://rkhullar.github.io/csci870/

[reu13]: https://rkhullar.github.io/csci870/images/report/reu-2013.png
[reu15]: https://rkhullar.github.io/csci870/images/report/reu-2015.png
[app1]: https://rkhullar.github.io/csci870/images/report/app-setup.png
[app2]: https://rkhullar.github.io/csci870/images/report/app-push.png
[erd]: https://rkhullar.github.io/csci870/images/report/erd.png
[pre]: https://rkhullar.github.io/csci870/images/report/preprocessing.png
[frw]: https://rkhullar.github.io/csci870/images/report/framework.png

[svm-linear]: http://www.scipy-lectures.org/_images/svm_kernel_linear.png
[svm-poly]: http://www.scipy-lectures.org/_images/svm_kernel_poly.png
[svm-radial]: http://www.scipy-lectures.org/_images/svm_kernel_rbf.png

[cntL]: https://rkhullar.github.io/csci870/figures/count/L.png
[cntT]: https://rkhullar.github.io/csci870/figures/count/T.png
[cntW]: https://rkhullar.github.io/csci870/figures/count/W.png
[cntLT]: https://rkhullar.github.io/csci870/figures/count/LT.png
[cntF]: https://rkhullar.github.io/csci870/figures/count/F.png

[distX]: https://rkhullar.github.io/csci870/figures/dist-x.png
[distWL]: https://rkhullar.github.io/csci870/figures/dist-WL.png
[decay]: https://rkhullar.github.io/csci870/figures/pdecay.png

[box1]: https://rkhullar.github.io/csci870/figures/boxplots/LT/EGGC_601_lab_at_10AM.png
[box2]: https://rkhullar.github.io/csci870/figures/boxplots/LT/EGGC_601_lab_at_12PM.png
[box3]: https://rkhullar.github.io/csci870/figures/boxplots/LT/EGGC_DL3_704_at_6PM.png
[box4]: https://rkhullar.github.io/csci870/figures/boxplots/LT/EGGC_DL3_704_at_7PM.png

[mtx1]: https://rkhullar.github.io/csci870/images/report/all_waps.png
[mtx2]: https://rkhullar.github.io/csci870/figures/matrices/W/01.png
[mtx3]: https://rkhullar.github.io/csci870/figures/matrices/WT/01.png
[mtx4]: https://rkhullar.github.io/csci870/figures/matrices/W/11.png
[mtx5]: https://rkhullar.github.io/csci870/figures/matrices/WT/11.png

HISTORY.md
## 2016-09-08
1. initialized database

## 2016-09-10
1. used "Let's Encrypt" to generate trusted certificates
2. setup https protocol on server

## 2016-09-11
1. created nydevteam gmail account.
2. learning how to send emails through python

## 2016-09-19
1. experimented with heroku for hosting the project
2. heroku's unpaid plan has a tight limit on data storage

## 2016-09-21
1. learned how to pass environment variables from apache vhost into flask

## 2016-09-22
1. finalized mail system

## 2016-09-25
1. converted multiple repos into one repo with multiple branches
2. initialized project page

## 2016-10-02
1. created python script to generate apache vhost config files

## 2016-10-03
1. initialized history

## 2016-10-09
1. initialized android sqlite database

## 2016-10-10
1. learned how to read request headers
2. created default location and actor in database to use for stage 1
3. added sql functions to aid data reconstruction from csv files

## 2016-10-11
1. learned how to make requests in python in order to test the api

## 2016-10-12
1. using jsonify for api responses
2. learned about request authorization

## 2016-10-14
1. plan out registration verification url
2. tried to read environment variables in postgres
3. decide to only support json media type for post methods

## 2016-10-15
1. initiate model database classes in python
2. work on person model for registration, login, and search
3. authorization decorator works
4. reimplemented error system
5. added json required decorator

## 2016-10-16
1. reimplemented authorization with decorators
2. now api endpoints can easily require password or token based authorization
3. reimplemented error system again to cover all error codes

## 2016-10-17 endpoint
3. learned how to print request responses properly

## 2016-10-18
1. modified auth decorator to add userid keyword arg
2. created endpoint to persist single scan
3. add dbo.admin and dbv.admin to database
4. add admin email and password to environment

## 2016-10-19
1. look into async task to call rest api from android

## 2016-10-20
1. add test activities to app
2. use async http client instead of async task
3. able to perform get request
4. able to add authentication to request
5. create model classes for scan and location

## 2016-10-22
1. learned how to add custom parameters to postgres database
2. added secret key to database

## 2016-10-23
1. updated new.user and fnd.user to include secret key
2. moving database inflation from sql to python
3. added base activity class for android

## 2016-10-24
1. generate verification url for user registration
2. restructured android app
3. app db create table statements successful
4. query, insert, and delete accomplished

## 2016-10-25

1. email verification endpoint
2. limited registration to edu domain
3. login on android app works
4. working on signup from app

## 2016-10-26
1. signup system works
2. mailer operates off main thread

## 2016-10-28
1. starting location csv inflation
2. modified new.location sql method to automatically add buildings
3. added api endpoint to fetch location list

## 2016-10-29
1. added a few locations to the list
2. changed android login system
3. authentication and fetch locations can be done in single request
4. flashed and rooted nexus 5 m4b30x
5. able to to debug sqlite on android by downloading directly
6. created fetch location system for app
7. fixed primary key typo in sqlite

## 2016-10-30
1. researching listview, spinner, and array adapter for android
2. tried converting wifi test scan to use listview instead of toast
3. wifi scan failed on marshmallow

## 2016-10-31
1. researching marshmallow permission system in order to fix wifi scan

## 2016-11-04
1. wifi scan working again, enabling wifi and calling disconnect before starting
2. listview for wifi scan works

## 2016-11-05
1. moved wifi scan receiver class outside of test
2. converted test buttons to home menu

## 2016-11-06
1. design user map for configuring and pushing scans
2. reimplemented home activity with fragments
3. update flask scan endpoint to return post data on success
4. created custom launcher icon

## 2016-11-07
1. test login and register endpoints for sql injection vulnerability

## 2016-11-08
1. reimplemented fragment classes to avoid extra files
2. add building, floor, and room queries to database adapter
3. working on spinners for scan configuration

## 2016-11-09
1. scan config location spinners works
2. updated database adapter floors method with 'distinct' and 'order by'
3. using number picker to select scan duration

## 2016-11-10
1. started implementing background wifi scans with intent service

## 2016-11-11
1. changed wifi scan service from intent service to normal service
2. using count down timer with interval one second

## 2016-11-12
1. working on getting service to run in foreground independent of activity
2. changed naming convention of home fragments
3. countdown timer in foreground service works

## 2016-11-13
1. added unix time api endpoint
2. added notification for when countdown timer complete
3. registered wifi scan receiver in manifest
4. able to perform multiple scans if permission already granted
5. learned how to use logcat system
6. permission request system works
7. changed scan table to use unix time integer instead of datetime
8. able to persist scans in wifi test activity

## 2016-11-14
1. updated android studio sdk platform
2. added method to fetch scan objects from sqlite

## 2016-11-15
1. added list view on scan push activity
2. updated logout method to preserve integrity; truncate all tables
3. implemented scan upload system with time synchronization

## 2016-11-16
1. compiled location list
2. push system on app works with minor bugs
3. fixed typo in database (wpa -> wap)
4. added database constraint to prevent repeatedly persisted scans
5. added endpoint to push multiple scan results
6. improved efficiency of app location download system

app.py
#!local/bin/python

"""
@author  :  Rajan Khullar
@created :  09/06/16
@updated :  12/04/16
"""

import decor as dec

from flask import Flask, jsonify, json, request, abort
from flask import render_template, send_file
from io import BytesIO
#from flask import Flask, Response, jsonify, json, request, abort, session, render_template, redirect, url_for

from error import apierror
from mail import send_thread_text

from core import core
from person import person
from location import location
from wap import wap
from scan import scan
from time import time

BASEURL = 'https://csci870.nydev.me'
BASEAPI = BASEURL + '/api'

app = Flask(__name__)
apierror.apply(app)

# Authorization Methods
pswd = dec.corify(person.pswd)
token = dec.corify(person.token)
admin = dec.corify(person.admin_login)

# User Methods
register = dec.corify(person.register)
verification = dec.corify(person.verification)
verify = dec.corify(person.verify)

# Scan Methods
fetch_locations = dec.corify(location.dump)
persist_scan = dec.corify(scan.persist)
persist_scans = dec.corify(scan.persistMany)
fetch_scans = dec.corify(scan.dump)


@app.route('/api/echo', methods=['GET', 'POST'])
def api_echo():
    if request.method == 'GET':
        return jsonify({'message': 'ok'})
    if request.method == 'POST':
        #return request.form['data']
        return request.json

@app.route('/api/time', methods=['GET'])
def api_time():
    resp = {'time': int(time())}
    return jsonify(resp)

@app.route('/api/authenticate')
@dec.auth(pswd)
def api_authenticate(userid):
    resp = {'message': 'ok'}
    return jsonify(resp)

@app.route('/api/admin')
@dec.auth(admin)
def api_admin(userid):
    resp = {'message': 'ok'}
    return jsonify(resp)

@app.route('/api/register', methods=['POST'])
@dec.json
def api_register():
    resp = {'message': 'not ok'}
    data = request.json
    t = register(**data)
    if(t):
        e = data['email']
        h = verification(t)
        u = '%s/verify/%s/%s' % (BASEURL, e, h)
        send_thread_text([e], 'nydev verify account', u)
        resp['message'] = 'ok'
    return jsonify(resp)

@app.route('/api/verify/<string:email>/<string:hash>', methods=['GET'])
def api_verify(email, hash):
    t = verify(email, hash)
    resp = {'message': 'ok'} if t else {'message': 'not ok'}
    return jsonify(resp)

@app.route('/verify/<string:email>/<string:hash>', methods=['GET'])
def web_verify(email, hash):
    t = verify(email, hash)
    if t:
        return render_template('verify.html')
    abort(400)

@app.route('/api/locations', methods=['GET'])
@dec.auth(pswd)
def api_locations(userid):
    n = 0
    resp = {'message': 'ok', 'building': [], 'floor': [], 'room':[]}
    for l in fetch_locations():
        resp['building'].append(l.building)
        resp['floor'].append(l.floor)
        resp['room'].append(l.room)
        n += 1
    resp['size'] = n
    return jsonify(resp)

@app.route('/api/scan', methods=['POST'])
@dec.auth(pswd)
@dec.json
def api_post_scan(userid):
    data = request.json
    data['userID'] = userid
    resp = {'message': 'not ok'}
    t = persist_scan(scan(**data))
    if t:
        resp['message'] = 'ok'
        for k in data:
            resp[k] = data[k]
    return jsonify(resp)

@app.route('/api/scans', methods=['POST'])
@dec.auth(pswd)
@dec.json
def api_post_scans(userid):
    data = request.json
    n  = int(data['size'])
    vt = data['uxt']
    vw = data['bssid']
    vl = data['level']
    vb = data['building']
    vf = data['floor']
    vr = data['room']
    l = []
    for i in range(n):
        x = scan(uxt=int(vt[i]), bssid=vw[i], level=int(vl[i]), building=vb[i], floor=int(vf[i]), room=vr[i])
        l.append(x)
    resp = persist_scans(userid, l)
    return jsonify(resp)

@app.route('/download/users', methods=['GET'])
@dec.auth(admin)
def download_users(userid):
    return download_csv('users.csv', person)

@app.route('/download/waps', methods=['GET'])
@dec.auth(admin)
def download_waps(userid):
    return download_csv('waps.csv', wap)

@app.route('/download/scans', methods=['GET'])
@dec.auth(admin)
def download_scans(userid):
    return download_csv('scans.csv', scan)

def download_csv(filename, kls):
    fn = dec.corify(kls.dump)
    buffer = BytesIO()
    buffer.write(kls().csvh().encode('utf-8'))
    for x in fn():
        buffer.write(x.csv().encode('utf-8'))
    buffer.seek(0)
    return send_file(buffer, as_attachment=True, attachment_filename=filename, mimetype='text/csv')

@app.route('/api/count/<mode>', methods=['GET'])
def api_count(mode):
    fn = dec.corify(scan.count)
    resp = fn(mode)
    return jsonify(resp)

if __name__ == '__main__':
    app.run(debug=True)
